lib/SILOptimizer/Utils/ConstExpr.cpp

```diff
@@ -0,0 +1,730 @@
+//===--- ConstExpr.cpp - Constant expression evaluator -------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2017 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+#define DEBUG_TYPE "ConstExpr"
+#include "swift/SILOptimizer/Utils/ConstExpr.h"
+#include "swift/AST/ProtocolConformance.h"
+#include "swift/AST/SubstitutionMap.h"
+#include "swift/Basic/Defer.h"
+#include "swift/Demangling/Demangle.h"
+#include "swift/SIL/ApplySite.h"
+#include "swift/SIL/FormalLinkage.h"
+#include "swift/SIL/SILBuilder.h"
+#include "swift/SIL/SILConstants.h"
+#include "swift/Serialization/SerializedSILLoader.h"
+#include "llvm/ADT/PointerEmbeddedInt.h"
+#include "llvm/Support/CommandLine.h"
+#include "llvm/Support/TrailingObjects.h"
+
+using namespace swift;
+
+static llvm::cl::opt<unsigned>
+    ConstExprLimit("constexpr-limit", llvm::cl::init(512),
+                   llvm::cl::desc("Number of instructions interpreted in a"
+                                  " constexpr function"));
+
+static llvm::Optional<SymbolicValue>
+evaluateAndCacheCall(SILFunction &fn, SubstitutionMap substitutionMap,
+                     ArrayRef<SymbolicValue> arguments,
+                     SmallVectorImpl<SymbolicValue> &results,
+                     unsigned &numInstEvaluated, ConstExprEvaluator &evaluator);
+
+// TODO: ConstantTracker in the performance inliner and the
+// ConstantFolding.h/cpp files should be subsumed by this, as this is a more
+// general framework.
+
+//===----------------------------------------------------------------------===//
+// ConstExprFunctionState implementation.
+//===----------------------------------------------------------------------===//
+
+namespace {
+/// This type represents the state of computed values within a function
+/// as evaluation happens.  A separate instance of this is made for each
+/// callee in a call chain to represent the constant values given the set of
+/// formal parameters that callee was invoked with.
+class ConstExprFunctionState {
+  /// This is the evaluator we put bump pointer allocated values into.
```

Comment:
@marcrasi Can we make this comment a bit more clear? The evaluator holds a reference to a bump allocator and also provides utility functions like getUnknown() etc.